packages/wallet/src/redux/protocols/ledger-top-up/reducer.ts

```diff
@@ -0,0 +1,288 @@
+import {
+  SharedData,
+  getChannel,
+  signAndStore,
+  queueMessage,
+  checkAndStore,
+  registerChannelToMonitor,
+} from '../../state';
+import * as states from './states';
+import { ProtocolStateWithSharedData, ProtocolReducer } from '..';
+import * as helpers from '../reducer-helpers';
+import { proposeNewConsensus, acceptConsensus } from '../../../domain';
+import { TwoPartyPlayerIndex } from '../../types';
+import { bigNumberify } from 'ethers/utils';
+import { theirAddress, getLastCommitment } from '../../channel-store';
+import { sendCommitmentReceived } from '../../../communication';
+import {
+  initialize as initializeDirectFunding,
+  directFundingStateReducer,
+} from '../direct-funding/reducer';
+import { LedgerTopUpAction } from './actions';
+import { directFundingRequested } from '../direct-funding/actions';
+import { isTerminal, isFailure, isSuccess } from '../direct-funding/states';
+import { addHex } from '../../../utils/hex-utils';
+export const LEDGER_TOP_UP_PROTOCOL_LOCATOR = 'LedgerTopUp';
+export function initialize(
+  processId: string,
+  channelId: string,
+  ledgerId: string,
+  proposedAllocation: string[],
+  proposedDestination: string[],
+  sharedData: SharedData,
+): ProtocolStateWithSharedData<states.LedgerTopUpState> {
+  const lastCommitment = getLatestCommitment(sharedData, ledgerId);
+
+  const newProtocolState = states.waitForPreTopUpLedgerUpdate({
+    processId,
+    ledgerId,
+    channelId,
+    proposedAllocation,
+    proposedDestination,
+  });
+
+  if (helpers.isFirstPlayer(ledgerId, sharedData)) {
+    const newAllocation = [
+      ...lastCommitment.allocation,
+      ...calculateTopUpAllocation(proposedAllocation, lastCommitment.allocation),
+    ];
+    const newDestination = [...lastCommitment.destination, ...proposedDestination];
```

Comment:
I don't think the rules enforce this, but I don't think multiple entries of the same participant is allowed as per the force-move protocol.

Reply:
I've created https://github.com/magmo/apps/issues/543 to address this in a separate PR. I thought I was being clever :(